### ISSUE TITLE
Terminal auto-scrolls with long input

### ISSUE THREAD

**Issue by fperez** [ipython/ipython#9390]
It seems after the introduction of the new prompt-toolkit machinery, the terminal automatically scrolls up, always leaving a few blank lines at the bottom.  This is different from how both the shell and previous IPython worked, and effectively wastes useful vertical space:

![image](https://cloud.githubusercontent.com/assets/57394/14486345/1fc15f70-0111-11e6-9c36-78fac134367c.png)

**Comment by takluyver** [comment]
It's reserving space so that when the tab completion dropdown appears, it doesn't force the terminal to jump up. I use tab completion a lot, so I think I like it that way, but it's a [parameter we pass](https://github.com/ipython/ipython/blob/de87d902bf3e17e90138cf5024fd7a77edee7b6a/IPython/terminal/ptshell.py#L249), so we could make it configurable if you'd rather have the extra space.

Ping @jonathanslenders for any extra thoughts.

**Comment by jonathanslenders** [comment]
This is a fair request.
I believed that it was the most user friendly to always reserve some space, rather than reserving the space for the menu when the first completion needs to be displayed. However, I think that this was mostly true for CLIs where the ``complete_while_typing`` parameter is True. (It means that you don't need to press tab, and completions are already displayed after the first key press.) Because in that case, the space is almost instantly required.

But if ``complete_while_typing`` is False, this may indeed seem like a little waste of space for people that don't use the tab key that often.

If nobody is against, I think I will add a ``min_space_for_menu`` parameter. If ``reserve_space_for_menu`` is zero, then this value will be used to reserve space at the point where the menu becomes visible. That way, people can configure it like they prefer. This will be for the next release of prompt-toolkit.

**Comment by takluyver** [comment]
> If nobody is against, I think I will add a min_space_for_menu parameter. If reserve_space_for_menu is zero, then this value will be used to reserve space at the point where the menu becomes visible.

Just to check I understand you, that parameter would effectively set the menu height? Would it only take effect when reserve_space is 0, or at any time? Is `reserve_space_for_menu` the only way at present to determine the menu size?

**Comment by jonathanslenders** [comment]
Both parameters will never set the menu height itself, but influence the vertical space used by the input region. The actual menu height still depends on the amount of completions and the position of the input. (The menu can appear both above and below the cursor.)

``reserve_space_for_menu`` ensures that the input region is at all times at least this height. (Except for the point where enter has been pressed, and we're rendering in the done/accept state.)

Now, we want a way to say that the input should be at least a certain height, but *only* if there are completions, and the completion menu is visible. I'm still not sure of the best API, but ``min_space_for_menu`` could be a way to do that. However, maybe a boolean could be better. I'm not yet sure about the parameter names. Ideas are welcome ;)

**Comment by takluyver** [comment]
Maybe instead of having two numeric parameters, it would be clearer to have the existing numeric parameter, plus a boolean parameter to determine whether that space is always reserved or only once completions are shown? Or maybe that should be folded into `complete_while_typing` for simplicity.

**Comment by jonathanslenders** [comment]
Actually, I really like the idea of folding it into ``complete_while_typing``. So, if that is True, always reserve the space, if False: only when a completion needs to be shown.

**Comment by jonathanslenders** [comment]
This should do it: https://github.com/jonathanslenders/python-prompt-toolkit/commit/22ddc1306da9234e5e9a277f96d1bfbc39a9fa38

**Comment by takluyver** [comment]
Great, thanks! Once that's released, we'll automatically get the user experience @fperez prefers, so I'll close this as there's nothing to do on our side.

**Comment by fperez** [comment]
OK, I'm happy to play with the behavior after the fix is released and will provide further feedback.  I was thinking about it earlier and wondering if providing an upward-deploying scroll area might not be a reasonable solution when at the bottom, but I don't know if you can reliably detect whether you're at the bottom or not...

Let's leave it closed for now. Please let me know when I should try it out...

**Comment by fperez** [comment]
And I failed to say, many thanks @jonathanslenders for the super-quick turnaround! PT is awesome, and I'm very happy we are getting all these great features thanks to your hard work :)

For a long time we wanted these kinds of features in the terminal, and tried to get someone to write a full curses client (to still be able to keep the simplicity of a purely line-oriented one).  But your work gives us the best of both worlds!

**Comment by jonathanslenders** [comment]
Thank you! :-)
